src/transform/src/nonnull_requirements.rs

```diff
@@ -195,16 +195,58 @@ impl NonNullRequirements {
                 expected_group_size: _,
             } => {
                 let mut new_columns = HashSet::new();
-                for column in columns {
-                    // No obvious requirements on aggregate columns.
-                    // A "non-empty" requirement, I guess?
-                    if column < group_key.len() {
-                        group_key[column].non_null_requirements(&mut new_columns);
+                let (group_key_columns, aggr_columns): (Vec<usize>, Vec<usize>) =
+                    columns.iter().partition(|c| **c < group_key.len());
+                for column in group_key_columns {
+                    group_key[column].non_null_requirements(&mut new_columns);
+                }
+
+                if !aggr_columns.is_empty() {
+                    // Intersect the non-null requirements for non-null aggregates
+                    // (one which its result is requested to be non-null)
+                    // that propagate non-null constraints (all of them but COUNT).
+                    // However, if there is a COUNT or any other aggregate we can still
+                    // propagate the non-null requirement as long as the requirements
+                    // of their parameters intersect with the requirements of a non-null
+                    // aggregate.
+                    let aggr_nonnull_req = aggregates
+                        .iter()
+                        .enumerate()
+                        .map(|(pos, aggr)| {
+                            let aggr_nonnull_flag = aggr.func.propagates_nonnull_constraint()
+                                && aggr_columns.contains(&(group_key.len() + pos));
+                            let mut aggr_nonnull_columns = HashSet::new();
+                            aggr.expr.non_null_requirements(&mut aggr_nonnull_columns);
+                            (aggr_nonnull_columns, aggr_nonnull_flag)
+                        })
+                        .collect::<Vec<_>>();
+                    let mut req_intersection: Option<HashSet<usize>> = None;
+                    for (nonnull_columns, nonnull_flag) in aggr_nonnull_req.iter() {
+                        if *nonnull_flag {
+                            if let Some(previous) = req_intersection {
+                                req_intersection = Some(
+                                    nonnull_columns.intersection(&previous).cloned().collect(),
+                                );
+                            } else {
+                                req_intersection = Some(nonnull_columns.clone());
+                            }
+                        } else if !aggr_nonnull_req
+                            .iter()
+                            .any(|(cols, flag)| *flag && cols.is_subset(&nonnull_columns))
```

Comment:
I don't understand this guard. I would understand `else` and then clearing `req_intersection`, but it looks like there is a more thoughtful test going on here, I just can't read what its intent is.

Reply:
Hopefully it is more clear after the last pushed commit.

If, for example, MAX(column) is requested to be non-null, the presence of COUNT(column), or any other aggregate on `column`, should not prevent the non-null requirement on `column` from being pushed down. This is actually the new case I was trying to address with the last commit before the one I've just made.